Customizable useragent (workaround for #127)

iamtakingiteasy opened the pull request:
PR is inteded to provide customizable useragent, in face of potential banwave of MJS+ users per #127

iamtakingiteasy commented:
For those unwilling to wait until this gets merged, temporary builds:
https://public.eientei.org/mjsp/Majsoul_Plus-2.0.1-windows-x64-Setup.exe
https://public.eientei.org/mjsp/Majsoul_Plus-2.0.1-linux-x64.tar.gz

iamtakingiteasy commented:
Added vanilla mode, builds are also updated.

Shaddox commented:
@iamtakingiteasy I would be very grateful for a x64 zip release.

Yesterday17 commented on i18n/zh-CN/manager.json:
```diff
@@ -17,6 +17,8 @@
   "useHttpServer": "使用 http 服务器（不存在证书错误）",
   "serverToPlay": "游戏服务器（0为国服，1为日服，2为美服）",
   "proxyUrl": "代理服务器地址（留空为不使用代理）",
+  "userAgent": "用户代理",
```
Just keep `User-Agent` here.

Yesterday17 commented on i18n/zh-CN/manager.json:
```diff
@@ -17,6 +17,8 @@
   "serverToPlay": "游戏服务器（0为国服，1为日服，2为美服）",
   "proxyUrl": "代理服务器地址（留空为不使用代理）",
+  "userAgent": "用户代理",
+  "vanillaMode": "Vanilla mode (no extensions)",
```
纯洁模式（停用所有扩展）

iamtakingiteasy commented:
@Yesterday17 thanks for the translations
@Shaddox builds are updated to contain last translation change and windows zip build is also added to original comment at https://github.com/MajsoulPlus/majsoul-plus/pull/128#issuecomment-873089143

Yesterday17 commented on src/windows/game.ts:
```diff
@@ -160,11 +160,12 @@ export function newGameWindow(id: number) {
 
   // 载入本地启动器
-  window.loadURL('file://' + path.join(__dirname, '../bin/main/index.html'))
+  window.loadURL('file://' + path.join(__dirname, '../bin/main/index.html'), {userAgent: UserConfigs.window.userAgent})
```
hmm, I think it's not necessary to set `User-Agent` for the window. Configuring `webContents.userAgent` should be enough.

Yesterday17 commented on i18n/zh-CN/manager.json:
```diff
@@ -17,6 +17,8 @@
   "serverToPlay": "游戏服务器（0为国服，1为日服，2为美服）",
   "proxyUrl": "代理服务器地址（留空为不使用代理）",
+  "userAgent": "User-Agent",
+  "vanillaMode": "纯洁模式（停用所有扩展）",
```
Sorry for typo, it should be `纯净模式` instead of `纯洁模式`.

Yesterday17 commented on i18n/zh-TW/manager.json:
```diff
@@ -21,6 +21,8 @@
   "useHttpServer": "\u4f7f\u7528 http \u670d\u52d9\u5668\uff08\u4e0d\u5b58\u5728\u8b49\u66f8\u932f\u8aa4\uff09",
   "serverToPlay": "\u904a\u6232\u4f3a\u670d\u5668\uff080\u70ba\u570b\u4f3a\uff0c1\u70ba\u65e5\u4f3a\uff0c2\u70ba\u7f8e\u4f3a\uff09",
   "proxyUrl": "\u4ee3\u7406\u4f3a\u670d\u5668\u5730\u5740\uff08\u7559\u7a7a\u70ba\u4e0d\u4f7f\u7528\u4ee3\u7406\uff09",
+  "userAgent": "\u7528\u6236\u4ee3\u7406",
```
`User-Agent` here too.

Yesterday17 commented on i18n/zh-TW/manager.json:
```diff
@@ -21,6 +21,8 @@
   "serverToPlay": "\u904a\u6232\u4f3a\u670d\u5668\uff080\u70ba\u570b\u4f3a\uff0c1\u70ba\u65e5\u4f3a\uff0c2\u70ba\u7f8e\u4f3a\uff09",
   "proxyUrl": "\u4ee3\u7406\u4f3a\u670d\u5668\u5730\u5740\uff08\u7559\u7a7a\u70ba\u4e0d\u4f7f\u7528\u4ee3\u7406\uff09",
+  "userAgent": "\u7528\u6236\u4ee3\u7406",
+  "vanillaMode": "Vanilla mode (no extensions)",
```
`\u7d14\u6de8\u6a21\u5f0f\uff08\u505c\u7528\u6240\u6709\u64f4\u5c55\uff09`

Yesterday17 commented on src/bin/main/mainLoader.ts:
```diff
@@ -105,7 +105,8 @@ function getLocalUrlWithParams(url: string) {
   console.log('[Majsoul Plus] Redirect Target:' + localUrl)
-  gameWindow.loadURL(localUrl)
+
+  gameWindow.loadURL(localUrl, {userAgent: navigator.userAgent})
```
I think it's not necessary to set `userAgent` here. See below.

Yesterday17 commented on src/bin/main/mainLoader.ts:
```diff
@@ -152,6 +153,8 @@ ipcRenderer.on(
       mainWindow.partition = partition
     }
+
+    mainWindow.useragent = navigator.userAgent
```
`mainWindow` is oriented to users. Setting `userAgent` here takes no effect to the game(and nobody would read `userAgent` here).

Yesterday17 reviewed (changes requested):
Looks nice. Preparing for merge after changes.

iamtakingiteasy commented:
Removed excessive user-agent specifications, correct i18n.

Additionally, removed start and user whatsoever of local server, when using vanilla mode.

Security considerations for non-vanilla mode:
- node-fetch is still susceptible to adding sec-fetch-* headers, by witch modified MJS+ client  may be identified
- even if network layer will get to the level of transparent proxy, being indistinguishable from regular client, this would not be measure enough to avoid detection, as at any time majsoul may return to portion of it's clients JS that would take snapshot of canvas bitmap and sending it to their servers for subsequent review, as well as inspect DOM for present alterations.

Both should not be an issue in current vanilla mode, however to implement truly undetectable majsoul client & mods, customizing electron's V8 to have untracable (e.g. requiring secret password/key to change/inspect) object properties, wrapping entirety of DOM API, as well as having mods to work as overlay to webContents instead of working inside of it, may be required.

This is quite a big undertaking (compiling electron alone may be quite a challenge, requiring 50G of space to build and maybe several days for build to complete), not to say all the mods have to be reworked with overlay approach in mind.

Yesterday17 commented:
hmm, I think we should validate how Yostar 'detects' modified clients first.

I used [this tool](https://github.com/Yesterday17/majsoul-source-analyze/tree/tool) to split majsoul code to smaller chunks and do some deobfuscations. It may work on EN server.

iamtakingiteasy commented:
Most likely their detection is not present in regular JS versions being returned to most users most of the time, but rather being returned to a percentage of the users when they feel like they want new banwave -- they could return to say 5,10,30,whatever percentage of version requests some new version, or simply return different contents for same version. It is very easy to implement, so analyzing their day-to-day code would likely yield nothing, catching banwave-altered code is tricky, would require big network of involved clients and/or telemetry and even if we'd manage to catch such code and counter it's current implementation directly - they easily can alter it in a way that our current counter does not expect. We of course could try catching it again, but always end up lagging behind, having some of the users banned. So we will never be able to reliably counter unknown code in any other way than executing it in contained and pristine environment, while having everything else (mods) as overlay or modified copy to it, that is being presented to a user.

That being said, vanilla mode of this PR should already enable those MJS+ users that just do not want to use their main browser for majsoul (e.g. for segregation and/or streaming purposes) to be out of harms way by emulating useragent of their preferred browser inside of MJS+.

iamtakingiteasy commented:
Apparently I forgot to clarify that retaining useragent override in `src/bin/main/mainLoader.ts#157` is required, as otherwise game window would inherit Majsoul useragent.

iamtakingiteasy commented:
Rebased with latest master; also made local builds using #151 changes, seems to resolve the sound effects overlapping.

https://public.eientei.org/mjsp/Majsoul_Plus-2.0.1-ua-1-windows-x64-Setup.exe
https://public.eientei.org/mjsp/Majsoul_Plus-2.0.1-ua-1-windows-x64.zip
https://public.eientei.org/mjsp/Majsoul_Plus-2.0.1-ua-1-linux-x64.tar.gz
